File: examples/python/postgres/postgres.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.postgres import PostgresConfig, PostgresResource
from mirage.types import PathSpec

load_dotenv(".env.development")

config = PostgresConfig(
    dsn=os.environ["POSTGRES_DSN"],
    max_read_rows=200,
    max_read_bytes=1024 * 1024,
)
resource = PostgresResource(config=config)


async def _run(ws, cmd):
    print(f"\n>>> {cmd}")
    r = await ws.execute(cmd)
    out = (await r.stdout_str()).strip()
    err = await r.stderr_str()
    if out:
        for line in out.splitlines()[:10]:
            print(f"  {line[:140]}")
        total = len(out.splitlines())
        if total > 10:
            print(f"  ... ({total} lines total)")
    if err:
        print(f"  [stderr] {err.strip()[:160]}")
    if not out and not err:
        print(f"  (empty, exit={r.exit_code})")
    return r


async def main():
    ws = Workspace({"/pg": resource}, mode=MountMode.READ)

    print("=" * 60)
    print("LISTING (ls / tree)")
    print("=" * 60)

    await _run(ws, "ls /pg/")
    await _run(ws, "ls /pg/public/")
    await _run(ws, "ls /pg/public/tables/")
    await _run(ws, "ls /pg/public/views/")
    await _run(ws, "tree -L 2 /pg/public/")

    print("\n" + "=" * 60)
    print("SYNTHETIC JSON (database.json + per-entity schema.json)")
    print("=" * 60)

    await _run(ws, "cat /pg/database.json | head -n 30")

    table = "context_info"
    sch = f"/pg/public/tables/{table}/schema.json"
    await _run(ws, f'cat "{sch}" | head -n 25')

    print("\n" + "=" * 60)
    print("STAT and METADATA")
    print("=" * 60)

    fp = f"/pg/public/tables/{table}/rows.jsonl"
    await _run(ws, f'stat "{fp}"')

    # chmod/chown/touch never hit Postgres: attrs land in the workspace
    # namespace (durable, snapshot-captured) and merge into
    # dispatch-level stat.
    print(f"=== metadata overlay on {fp} ===")
    meta_res = await ws.execute(f'chmod 640 "{fp}" && chown 500:dev "{fp}"'
                                f' && touch -t 202601021530 "{fp}"')
    print(f"  chmod/chown/touch exit={meta_res.exit_code}")
    try:
        meta_st, _ = await ws.dispatch("stat", PathSpec.from_str_path(fp))
        print(
            f"  dispatch stat: mode={oct(meta_st.mode)[2:]} uid={meta_st.uid} "
            f"gid={meta_st.gid} mtime={meta_st.modified}")
    except FileNotFoundError:
        print("  dispatch stat: target missing in this environment")

    print("\n" + "=" * 60)
    print("HEAD / TAIL / WC (predicate pushdown to SQL)")
    print("=" * 60)

    await _run(ws, f'head -n 3 "{fp}"')
    await _run(ws, f'tail -n 3 "{fp}"')
    await _run(ws, f'wc -l "{fp}"')

    print("\n" + "=" * 60)
    print("CAT with SIZE GUARD")
    print("=" * 60)
    print("(this should error: table has more rows than max_read_rows=200)")

    await _run(ws, f'cat "{fp}"')

    print("\n" + "=" * 60)
    print("GREP at different scopes (ILIKE pushdown)")
    print("=" * 60)

    await _run(ws, f'grep system "{fp}"')
    await _run(ws, f'grep -c system "{fp}"')
    await _run(ws, 'grep system /pg/public/tables/')
    await _run(ws, 'grep system /pg/public/')

    print("\n" + "=" * 60)
    print("RG at schema scope")
    print("=" * 60)

    await _run(ws, "rg system /pg/public/")

    print("\n" + "=" * 60)
    print("JQ over rows.jsonl (slice with head first to stay under guard)")
    print("=" * 60)

    await _run(ws, f'head -n 20 "{fp}" | jq -r ".[] | .id"')
    await _run(ws, f'head -n 20 "{fp}" | jq -r ".[] | .file_name"')

    print("\n" + "=" * 60)
    print("FIND")
    print("=" * 60)

    await _run(ws, 'find /pg/public/tables/ -name rows.jsonl')

    print("\n" + "=" * 60)
    print("CD + relative paths")
    print("=" * 60)

    await ws.execute("cd /pg/public/tables")
    await _run(ws, "pwd")
    await _run(ws, "ls")

    await resource.accessor.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/postgres/postgres.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import dotenv from 'dotenv'
import { MountMode, PostgresResource, Workspace, type FileStat } from '@struktoai/mirage-node'

dotenv.config({ path: '.env.development' })

const dsn = process.env.POSTGRES_DSN
if (dsn === undefined) {
  console.error('POSTGRES_DSN missing in .env.development')
  process.exit(1)
}

const resource = new PostgresResource({ dsn })
const ws = new Workspace({ '/pg/': resource }, { mode: MountMode.READ })

const DEC = new TextDecoder()

async function run(label: string, cmd: string): Promise<void> {
  console.log(`\n=== ${label} ===`)
  const r = await ws.execute(cmd)
  if (r.exitCode !== 0) {
    console.log(`(exit=${String(r.exitCode)})`)
    if (r.stderr.byteLength > 0) console.log(DEC.decode(r.stderr))
    if (r.stdout.byteLength > 0) console.log(DEC.decode(r.stdout))
    return
  }
  process.stdout.write(DEC.decode(r.stdout))
  if (!DEC.decode(r.stdout).endsWith('\n')) process.stdout.write('\n')
}

try {
  await run('ls /pg', 'ls /pg')

  await run('cat /pg/database.json (head)', 'head -n 20 /pg/database.json')

  const dirOut = await ws.execute('ls /pg/public/tables')
  const tables = DEC.decode(dirOut.stdout).split('\n').filter((s) => s.length > 0)
  if (tables.length === 0) {
    console.log('\nno tables in public schema; stopping')
  } else {
    const target = tables[0]!
    const dir = `/pg/public/tables/${target}`

    await run(`ls ${dir}`, `ls ${dir}`)
    await run('stat schema.json', `stat ${dir}/schema.json`)
    await run('stat rows.jsonl', `stat ${dir}/rows.jsonl`)


    // chmod/chown/touch never hit Postgres: attrs land in the
    // workspace namespace (durable, snapshot-captured) and merge into
    // dispatch-level stat.
    console.log(`=== metadata overlay on ${dir}/rows.jsonl ===`)
    const metaRes = await ws.execute(
      `chmod 640 "${dir}/rows.jsonl" && chown 500:dev "${dir}/rows.jsonl" && touch -t 202601021530 "${dir}/rows.jsonl"`,
    )
    console.log(`  chmod/chown/touch exit=${String(metaRes.exitCode)}`)
    try {
      const metaSt = (await ws.dispatch('stat', `${dir}/rows.jsonl`)) as FileStat
      const metaMode = metaSt.mode !== null ? metaSt.mode.toString(8) : '-'
      console.log(
        `  dispatch stat: mode=${metaMode} uid=${String(metaSt.uid)} gid=${String(metaSt.gid)} mtime=${String(metaSt.modified)}`,
      )
    } catch {
      console.log('  dispatch stat: target missing in this environment')
    }
    await run('cat schema.json', `cat ${dir}/schema.json`)
    await run('head -n 3 rows.jsonl', `head -n 3 ${dir}/rows.jsonl`)
    await run('tail -n 2 rows.jsonl', `tail -n 2 ${dir}/rows.jsonl`)
    await run('wc -l rows.jsonl', `wc -l ${dir}/rows.jsonl`)
    await run('grep ":" /pg/public/tables (top-level pushdown)', `grep -c ":" ${dir}/rows.jsonl`)
    await run('jq .name schema.json', `jq ".name" ${dir}/schema.json`)
    await run('find /pg/public -maxdepth 2', 'find /pg/public -maxdepth 2')
    await run('tree -L 2 /pg/public', 'tree -L 2 /pg/public')

    await run(
      'cat rows.jsonl (size guard fires for huge tables)',
      `cat ${dir}/rows.jsonl | wc -l`,
    )
  }

  console.log('\n=== ops summary ===')
  // workspace exposes ops counts via internal observer; we just confirm we got here cleanly.
  console.log('all commands completed')
} finally {
  await ws.close()
  await resource.close()
}
